examples/common/chip-app-server/QRCodeUtil.cpp

```diff
@@ -75,3 +88,44 @@ CHIP_ERROR GetQRCode(uint32_t & setupPinCode, std::string & QRCode, chip::Rendez
 exit:
     return err;
 }
+
+static inline bool isCharUnreservedInRfc3986(const char c)
+{
+    return isalpha(c) || isdigit(c) || (strchr(specialCharsUnreservedInRfc3986, c) != nullptr);
+}
+
+CHIP_ERROR EncodeQRCodeToUrl(const char * QRCode, size_t len, char * url, size_t maxSize)
+{
+    const char upperHexDigits[] = "0123456789ABCDEF";
+    CHIP_ERROR err              = CHIP_NO_ERROR;
+    size_t i = 0, j = 0;
+
+    VerifyOrExit((QRCode && url) != NULL, err = CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrExit(len <= maxSize, err = CHIP_ERROR_INVALID_ARGUMENT);
+
+    for (i = 0; i < len; ++i)
+    {
+        unsigned char c = QRCode[i];
+        if (isCharUnreservedInRfc3986(c))
+        {
+
+            VerifyOrExit((j + 1) < maxSize, err = CHIP_ERROR_BUFFER_TOO_SMALL);
+
+            url[j++] = c;
+        }
+        else
+        {
+
+            VerifyOrExit((j + 3) < maxSize, err = CHIP_ERROR_BUFFER_TOO_SMALL);
+
+            url[j++] = '%';
+            url[j++] = upperHexDigits[c >> 4];
```

Comment:
I would really prefer it if this were `(c & 0xf0) >> 4` just to be safe.  Nothing says `char` can't represent values larger then 255, and then you'd potentially index out of the array here.

Yes, I know you just moved the code.